### DISCUSSION TITLE
Function that takes transaction or connection as argument

### DISCUSSION THREAD

**Discussion by kulak** [blackbeam/mysql_async#310] [Q&A] [answered]
It is nice to be able to write code that's independent of current status of connection or transaction and thus function could take executor that is either transaction or connection.

Can it be done with this library?

It seems that I might be able to ask for connection and then pass transaction that gets converted to underlying connection, but I am not sure.

**Comment by blackbeam** [comment, marked as answer]
Hi. Please look at the following code snippet:

```rust
pub async fn archive(conn: &mut impl Queryable, id: i32) -> crate::Result<bool> {
                  // ^^^^^^^^^^^^^^^^^^^^^^^^^ this works for both &mut Conn and &mut Transaction
    const QUERY: &str = "UPDATE `ReadingRooms` \
        SET `is_archived` = NOT `is_archived` WHERE `id` = ?";
    let affected_rows = conn.exec_iter(QUERY, (id,)).await?.affected_rows();
    Ok(affected_rows > 0)
}
```

**Reply by kulak** [reply]
Thank you, I think it is exactly what I was looking for.  I saw Queryable, but I was not sure if it was the intent.

**Reply by kulak** [reply]
The problem with the above example, is that it does not seem to work with `last_insert_id`, as that one is defined on DB connection.

**Reply by blackbeam** [reply]
> The problem with the above example, is that it does not seem to work with `last_insert_id`, as that one is defined on DB connection.

Looks like you are looking for [`QueryResult::last_insert_id`](https://docs.rs/mysql_async/latest/mysql_async/struct.QueryResult.html#method.last_insert_id)
